src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilteringJsonGenerator.java

```diff
@@ -0,0 +1,432 @@
+import java.util.List;
+import java.util.Queue;
+
+/**
+ * A FilteringJsonGenerator uses antpath-like filters to include/exclude fields when writing XContent streams.
+ *
+ * When writing a XContent stream, this class instanciates a FilterContext instance for each field (or property)
+ * that must be generated. This filter context is used to check if the field/property must be
+ * written according to the current list of XContentFilter filters.
+ */
+public class FilteringJsonGenerator extends BaseJsonGenerator {
+
+    /**
+     * List of previous contexts
+     * (MAX_CONTEXTS contexts are kept around in order in order to be reused)
+     */
+    private Queue<FilterContext> contexts = new ArrayDeque<>();
+    private static final int MAX_CONTEXTS = 10;
+
+    /**
+     * Current filter context
+     */
+    private FilterContext context;
+
+    public FilteringJsonGenerator(JsonGenerator generator, String[] filters) {
+        super(generator);
+
+        // Convert each antpath-like filter into a XContentFilter
+        ImmutableList.Builder<XContentFilter> builder = ImmutableList.builder();
+        if (filters != null) {
+            for (String filter : filters) {
+                XContentFilter matcher = parse(filter);
+                if (matcher != null) {
+                    builder.add(matcher);
+                }
+            }
+        }
+
+        // Creates a root context that matches all filtering rules
+        this.context = acquire(null, null, builder.build());
+    }
+
+    /**
+     * Parse a filter and returns a XContentFilter. The filter "nodes.*.host" is converted
+     * into 3 ordered XContentFilter instances: nodes -> * -> host -> null
+     */
+    private static XContentFilter parse(String filter) {
+        String[] fields = Strings.delimitedListToStringArray(filter, ".");
+
+        XContentFilter next = null;
+        for (int i = fields.length; i > 0; i--) {
+            next = new XContentFilter(fields[i - 1], next);
+        }
+        return next;
+    }
+
+    /**
+     * Get a new context instance (and reset it if needed)
+     */
+    private FilterContext acquire(String property, FilterContext parent) {
+        FilterContext ctx = contexts.poll();
+        if (ctx == null) {
+            ctx = new FilterContext(property, parent);
+        } else {
+            ctx.reset(property, parent);
+        }
+        return ctx;
+    }
+
+    /**
+     * Get a new context instance (and reset it if needed)
+     */
+    private FilterContext acquire(String property, FilterContext context, List<XContentFilter> matchings) {
+        FilterContext ctx = acquire(property, context);
+        if (matchings != null) {
+            for (XContentFilter matching : matchings) {
+                ctx.addMatching(matching);
+            }
+        }
+        return ctx;
+    }
+
+    /**
+     * Adds a context instance to the pool in order to reuse it if needed
+     */
+    private void release(FilterContext ctx) {
```

Comment:
I renamed `acquire` into `get` and `release` into `put`, does it sound better?

Right, we only keep a limited number of contexts (10) around in order to reuse them. In my tests this list did not exceed 3 instances.